File: tests/src/com/bugsnag/android/test/BeforeNotifyTest.java
package com.bugsnag.android;

import com.bugsnag.android.BeforeNotify;
import com.bugsnag.android.Error;

public class BeforeNotifyTest extends BugsnagTestCase {
    BeforeNotify beforeNotify = new BeforeNotify() {
        @Override
        public boolean run(Error error) {
            return true;
        }
    };

    BeforeNotify beforeNotifySkip = new BeforeNotify() {
        @Override
        public boolean run(Error error) {
            return false;
        }
    };

    @Override
    protected void setUp() {

    }

    public void testRunModifiesError() {
        BeforeNotify beforeNotify = new BeforeNotify() {
            @Override
            public boolean run(Error error) {
                error.setContext("new-context");
                return false;
            }
        };

        Configuration config = new Configuration("api-key");
        Error error = new Error(config, new RuntimeException("Test"));
        beforeNotify.run(error);

        assertEquals(error.getContext(), "new-context");
    }

    public void testRunAll() {
        // Any false beforeNotify should return false
        Configuration config = new Configuration("api-key");
        config.addBeforeNotify(beforeNotify);
        config.addBeforeNotify(beforeNotifySkip);

        Error error = new Error(config, new RuntimeException("Test"));
        assertFalse(BeforeNotify.runAll(config.beforeNotifyTasks, error));
    }
}

File: bugsnag-android-core/src/androidTest/java/com/bugsnag/android/BeforeNotifyTest.kt
package com.bugsnag.android

import org.junit.Assert.assertEquals

import androidx.test.filters.SmallTest
import com.bugsnag.android.BugsnagTestUtils.generateSessionTracker

import org.junit.Test

@SmallTest
class BeforeNotifyTest {

    private val config = BugsnagTestUtils.generateImmutableConfig()

    @Test
    fun testRunModifiesError() {
        val context = "new-context"

        val beforeNotify = BeforeNotify {
            it.context = context
            false
        }

        val error = Error.Builder(config, RuntimeException("Test"), generateSessionTracker(),
            Thread.currentThread(), false, MetaData()).build()
        beforeNotify.run(error)
        assertEquals(context, error.context)
    }
}
